### PR TITLE
pointer-confinement

### PR DESCRIPTION
**Pull request by AlanGriffiths**
Adds test infrastructure for pointer_constraints and relative_pointer

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
The PR only contains a few initial tests, but allows for feedback on the approach. (The Mir spike of these protocols doesn't pass even these example tests yet.)

File: CMakeLists.txt [hunk 1]
```diff
@@ -170,6 +176,7 @@ set(
   ${PROTOCOL_SOURCES}
 
   tests/test_bad_buffer.cpp
+        tests/pointer_constraints.cpp
```
Comment (wmww) [inline CMakeLists.txt:179]: Indentation

File: src/pointer_constraints_unstable_v1.cpp [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+    return manager;
+}
+
+wlcs::ZwpConfinedPointerV1::ZwpConfinedPointerV1(
+    ZwpPointerConstraintsV1& manager,
+    wl_surface* surface,
+    wl_pointer* pointer,
+    wl_region* region,
+    uint32_t lifetime) :
+    relative_pointer{zwp_pointer_constraints_v1_confine_pointer(manager, surface, pointer, region, lifetime)},
```
Comment (wmww) [inline src/pointer_constraints_unstable_v1.cpp:42]: Don't we generally format this like
```
    uint32_t lifetime)
    : relative_pointer...
```
?

File: src/pointer_constraints_unstable_v1.cpp [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+    };
+
+
+wlcs::ZwpLockedPointerV1::ZwpLockedPointerV1(
+    ZwpPointerConstraintsV1& manager,
+    wl_surface* surface,
+    wl_pointer* pointer,
+    wl_region* region,
+    uint32_t lifetime) :
+    relative_pointer{zwp_pointer_constraints_v1_lock_pointer(manager, surface, pointer, region, lifetime)},
```
Comment (wmww) [inline src/pointer_constraints_unstable_v1.cpp:72]: Formatting again

File: tests/pointer_constraints.cpp [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    EXPECT_CALL(locked_ptr, locked()).Times(1);
+
+    client.roundtrip();
+}
+
+TEST_F(PointerConstraints, locked_pointer_does_not_move)
+{
+    auto const initial_cursor_position = client.pointer_position();
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, nw_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+    EXPECT_CALL(locked_ptr, locked()).Times(AnyNumber());
```
Comment (wmww) [inline tests/pointer_constraints.cpp:86]: Perhaps add `<< "Precondition failed"` to make it clear this isn't the main thing this test is testing

File: tests/pointer_constraints.cpp [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    client.roundtrip();
+}
+
+TEST_F(PointerConstraints, when_cursor_moves_onto_surface_locked_pointer_gets_locked_notification)
+{
+    ZwpLockedPointerV1 locked_ptr{pointer_constraints, se_surface, pointer, nullptr, ZWP_POINTER_CONSTRAINTS_V1_LIFETIME_ONESHOT};
+
+    EXPECT_CALL(locked_ptr, locked()).Times(1);
+
+    cursor.move_to(se_middle_x, se_middle_y);
```
Comment (wmww) [inline tests/pointer_constraints.cpp:110]: This seems like an implementation detail. In fact, an annoying amount of stuff about how this protocol behaves are implementation-dependent. I'm find with the more basic stuff (locking on a focused surface working immediately), but this is likely to actually be different from compositor to compositor. For example, some might require clicking on the surface before locking.

File: src/pointer_constraints_unstable_v1.cpp (hunk 2 above)
Comment (AlanGriffiths) [reply src/pointer_constraints_unstable_v1.cpp:42]: No, we normally format like this PR. What you show is [option 3](https://mir-server.io/doc/cppguide/?showone=Constructor_Initializer_Lists#Constructor_Initializer_Lists).

File: tests/pointer_constraints.cpp (hunk 4 above)
Comment (AlanGriffiths) [reply tests/pointer_constraints.cpp:86]: It isn't going to fail. I should, perhaps, use `NiceMock<>`

File: tests/pointer_constraints.cpp (hunk 5 above)
Comment (AlanGriffiths) [reply tests/pointer_constraints.cpp:110]: True, the trouble is we don't (yet) have ways to express the requirement to the compositor. I.e. `the_server().capture(cursor, on(se_surface))`.

File: src/pointer_constraints_unstable_v1.cpp (hunk 2 above)
Comment (wmww) [reply src/pointer_constraints_unstable_v1.cpp:42]: Oh, I forgot we had multiple acceptable options there

File: tests/pointer_constraints.cpp (hunk 5 above)
Comment (wmww) [reply tests/pointer_constraints.cpp:110]: Fair. Maybe add a click here though? I'm not convinced it's desirable for any compositor to capture the pointer on mouse over. It should at least need to be focused, no?

Comment (AlanGriffiths) [reply tests/pointer_constraints.cpp:110]: Interestingly, adding the click caused my local iteration of the prototype implementation to pass the test.

**Review by wmww — APPROVED** [review]
LGTM
bors r+

**Pull request merged**